File: archive/canvas-api/src/main/java/com/instructure/canvasapi/model/StreamItem.java
/*
 * Copyright (C) 2017 - present Instructure, Inc.
 *
 *     Licensed under the Apache License, Version 2.0 (the "License");
 *     you may not use this file except in compliance with the License.
 *     You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *     Unless required by applicable law or agreed to in writing, software
 *     distributed under the License is distributed on an "AS IS" BASIS,
 *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *     See the License for the specific language governing permissions and
 *     limitations under the License.
 *
 */

package com.instructure.canvasapi.model;

import android.content.Context;
import android.os.Parcel;

import com.google.gson.annotations.SerializedName;
import com.instructure.canvasapi.R;
import com.instructure.canvasapi.utilities.APIHelpers;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Map;

public class StreamItem extends CanvasModel<StreamItem> {

    public enum Type { DISCUSSION_TOPIC, SUBMISSION, ANNOUNCEMENT, CONVERSATION, MESSAGE, CONFERENCE, COLLABORATION, COLLECTION_ITEM, UNKNOWN, NOT_SET;
        public static boolean isDiscussionTopic(StreamItem streamItem) {return streamItem.getType() == DISCUSSION_TOPIC;}
        public static boolean isSubmission(StreamItem streamItem) {return streamItem.getType() == SUBMISSION;}
        public static boolean isAnnouncement(StreamItem streamItem) {return streamItem.getType() == ANNOUNCEMENT;}
        public static boolean isConversation(StreamItem streamItem) {return streamItem.getType() == CONVERSATION;}
        public static boolean isMessage(StreamItem streamItem) {return streamItem.getType() == MESSAGE;}
        public static boolean isConference(StreamItem streamItem) {return streamItem.getType() == CONFERENCE;}
        public static boolean isCollaboration(StreamItem streamItem) {return streamItem.getType() == COLLABORATION;}
        public static boolean isCollectionItem(StreamItem streamItem) {return streamItem.getType() == COLLECTION_ITEM;}
        public static boolean isUnknown(StreamItem streamItem) {return streamItem.getType() == UNKNOWN;}
        public static boolean isNotSet(StreamItem streamItem) {return streamItem.getType() == NOT_SET;}
    }

    // general info returned by the API
    private String updated_at;
    private long id;
    private String title;
    private String message;
    private String type;
    private String context_type;
    private boolean read_state;
    private String url;
    private String html_url;
    private long course_id = -1;
    private long group_id = -1;
    private long assignment_id = -1;

    // message type, which is not a conversation, but stream notifications
    private long message_id;
    private String notification_category;

    // conversation
    private long conversation_id;
    @SerializedName("private")
    private boolean isPrivate;
    private int participant_count;

    // discussionTopic or announcement
    private long discussion_topic_id = -1;
    private long announcement_id;
    private int total_root_discussion_entries;
    private boolean require_initial_post;
    private boolean user_has_posted;
    private List<DiscussionEntry> root_discussion_entries = new ArrayList<DiscussionEntry>();

    // submission
    private int attempt;
    private String body;
    private String grade;
    private boolean grade_matches_current_submission;
    private String graded_at;
    private long grader_id;
    private double score = -1.0;
    private String submission_type;
    private String submitted_at;
    private String workflow_state;
    private boolean late;
    private String preview_url;
    private List<SubmissionComment> submission_comments = new ArrayList<SubmissionComment>();
    private CanvasContext canvasContext;
    private Assignment assignment;
    private long user_id;
    private User user;

    // helper fields
    private Type enumType = Type.NOT_SET;
    private CanvasContext.Type canvasContextType = CanvasContext.Type.USER;
    private boolean hasSetContextType = false;
    private Date updatedAtDate;
    private Date gradedAtDate;
    private Date submittedAtDate;
    private Conversation conversation;
    private boolean isChecked;

    ///////////////////////////////////////////////////////////////////////////
    // Required Overrides
    ///////////////////////////////////////////////////////////////////////////

    // We want opposite of natural sorting order of date since we want the newest one to come first
    @Override
    public Date getComparisonDate() {
        return  getUpdatedAtDate();
    }

    @Override
    public String getComparisonString() {
        return title;
    }

    ///////////////////////////////////////////////////////////////////////////
    // Getters
    ///////////////////////////////////////////////////////////////////////////

    public Date getUpdatedAtDate() {
        if (updatedAtDate == null) {
            updatedAtDate = APIHelpers.stringToDate(updated_at);
        }
        return updatedAtDate;
    }
    public long getId() {
        return id;
    }
    public String getTitle(Context context) {
        if (title == null && getType() == Type.CONVERSATION) {
            title = context.getString(R.string.Message);
        }
        return title;
    }
    public String getMessage(Context context) {
        if (message == null) {
            message = createMessage(context);
        }
        return message;
    }
    public Type getType() {
        if (enumType == Type.NOT_SET) {
            enumType = typeFromString(type);
        }
        return enumType;
    }
    public CanvasContext.Type getContextType() {
        if (!hasSetContextType) {
            if (context_type != null && (context_type.toLowerCase().equals("course") || course_id > 0)) {
                canvasContextType = CanvasContext.Type.COURSE;
            } else if (context_type != null && (context_type.toLowerCase().equals("group") || group_id > 0)){
                canvasContextType = CanvasContext.Type.GROUP;
            }
            hasSetContextType = true;
        }

        return canvasContextType;
    }
    public boolean isReadState() {
        return read_state;
    }

    //helper method to show that the stream item has been read without having to reload all the data.
    //this method does not get the data from the server, so make sure item is actually read.
    public void setReadState(boolean readState) {
        read_state = readState;
    }
    public String getUrl() {
        return url;
    }
    public String getHtmlUrl() {
        return html_url;
    }
    public long getCourseId() {
        if (getContextType() == CanvasContext.Type.COURSE && course_id == -1) {
            course_id = parseCourseId();
        }
        return course_id;
    }

    public long getGroupId() {
        if (getContextType() == CanvasContext.Type.GROUP && group_id == -1) {
            group_id = parseGroupId();
        }
        return group_id;
    }
    public long getAssignmentId() {
        if (getContextType() == CanvasContext.Type.COURSE && assignment_id == -1) {
            assignment_id = parseAssignmentId();
        }
        return assignment_id;
    }
    public long getMessageId() {
        return message_id;
    }
    public String getNotificationCategory() {
        return notification_category;
    }
    public long getConversationId() {
        return conversation_id;
    }
    public boolean isPrivate() {
        return isPrivate;
    }
    public int getParticipantCount() {
        return participant_count;
    }
    public long getDiscussionTopicId() {
        if (discussion_topic_id == -1) {
            return announcement_id;
        }
        return discussion_topic_id;
    }
    public int getTotalRootDiscussionEntries() {
        return total_root_discussion_entries;
    }
    public boolean requiresInitialPost() {
        return require_initial_post;
    }
    public boolean userHasPosted() {
        return user_has_posted;
    }
    public List<DiscussionEntry> getRootDiscussionEntries() {
        return root_discussion_entries;
    }
    public int getAttempt() {
        return attempt;
    }
    public String getBody() {
        return body;
    }

    public String getGrade() {
        return grade;
    }

    public boolean gradeMatchesCurrentSubmission() {
        return grade_matches_current_submission;
    }
    public Date getGradedAt() {
        if (gradedAtDate == null) {
            gradedAtDate = APIHelpers.stringToDate(graded_at);
        }
        return gradedAtDate;
    }
    public long getGraderId() {
        return grader_id;
    }
    public double getScore() {
        return score;
    }
    public String getSubmissionType() {
        return submission_type;
    }
    public Date getSubmittedAt() {
        if (submittedAtDate == null) {
            submittedAtDate = APIHelpers.stringToDate(submitted_at);
        }
        return submittedAtDate;
    }

    public String getWorkflowState() {
        return workflow_state;
    }

    public boolean isLate() {
        return late;
    }

    public String getPreviewUrl() {
        return preview_url;
    }

    public List<SubmissionComment> getSubmissionComments() {
        return submission_comments;
    }

    public CanvasContext getCanvasContext() {
        return canvasContext;
    }

    public Assignment getAssignment() {
        return assignment;
    }

    public long getUser_id() {
        return user_id;
    }

    public User getUser() {
        return user;
    }
    public Conversation getConversation() {
        return conversation;
    }
    public boolean isChecked() {
        return isChecked;
    }
    public void setChecked(boolean isChecked) {
        this.isChecked = isChecked;
    }
    ///////////////////////////////////////////////////////////////////////////
    // Helpers
    ///////////////////////////////////////////////////////////////////////////

    private Type typeFromString(String type) {
        if(type.toLowerCase().equals("conversation")) {
            return Type.CONVERSATION;
        } else if(type.toLowerCase().equals("submission")) {
            return Type.SUBMISSION;
        } else if(type.toLowerCase().equals("discussiontopic")) {
            return Type.DISCUSSION_TOPIC;
        } else if (type.toLowerCase().equals("announcement")){
            return Type.ANNOUNCEMENT;
        } else if(type.toLowerCase().equals("message")) {
            return Type.MESSAGE;
        } else if(type.toLowerCase().equals("conference")) {
            return Type.CONFERENCE;
        } else if(type.toLowerCase().equals("webconference")) {
            return Type.CONFERENCE;
        } else if(type.toLowerCase().equals("collaboration")) {
            return Type.COLLABORATION;
        } else if(type.toLowerCase().equals("collectionitem")) {
            return Type.COLLECTION_ITEM;
        }
        return Type.UNKNOWN;
    }

    public void setConversation(Context context, Conversation conversation, long myUserId, String monologueDefault) {

        if(APIHelpers.paramIsNull(context, conversation,monologueDefault)){
            return;
        }

        this.conversation = conversation;
        title = conversation.getMessageTitle(context, myUserId,monologueDefault);
        message = createMessage(context);
    }

    public void setCanvasContextFromMap(Map<Long, Course> courseMap, Map<Long, Group> groupMap) {
        if (getContextType() == CanvasContext.Type.COURSE) {
            canvasContext = courseMap.get(getCourseId());
        } else {
            canvasContext = groupMap.get(getGroupId());
        }
    }

    private String createMessage(Context context) {
        switch (getType()) {
            case CONVERSATION:
                if (conversation == null) {
                    return context.getString(R.string.Loading);
                } else if (conversation.getLastMessagePreview() == null) {
                    return "";
                }
                return conversation.getLastMessagePreview();
            case SUBMISSION:
                //get comments from assignment
                String comment = null;
                if (submission_comments.size()> 0) {
                    comment = submission_comments.get(submission_comments.size() - 1).getComment();
                }
                //set it to the last comment if it's not null
                if(comment != null && !comment.equals("null") && score != -1.0) {
                    return (":" + score + " " + comment);
                }
                else if((comment == null || comment.equals("null")) && score != -1.0){
                    return (":" + score);
                }
                else if(comment != null && !comment.equals("null") && score == -1.0) {
                    return (comment);
                }
                break;
            case DISCUSSION_TOPIC :
                //if it's a discussionTopic, get the last entry for the message.
                if (root_discussion_entries.size() > 0) {
                    return root_discussion_entries.get(root_discussion_entries.size() - 1).getMessage(context.getString(R.string.Deleted));
                }
                break;
            default:
                break;
        }

        if (message == null) {
            return "";
        }
        return message;
    }

    private long parseAssignmentId() {
        //get the assignment from the url
        if(html_url.length() > 0 && !html_url.equals("null")) {
            String searchFor = "assignments/";
            int start = html_url.indexOf(searchFor);
            if (start == -1) {
                return 0;
            }
            start += searchFor.length();
            int end = html_url.indexOf("/", start);
            //in some urls the assignmentID might be the last thing so there wouldn't be a final /
            if(end == -1) {
                end = html_url.length();
            }
            String assignmentId = html_url.substring(start,end);

            return Long.parseLong(assignmentId);
        }
        return 0;
    }

    private long parseCourseId() {
        if(html_url.length() > 0 && !html_url.equals("null")) {
            String searchFor = "courses/";
            int start = html_url.indexOf(searchFor);
            if (start == -1) {
                return 0;
            }
            start += searchFor.length();
            int end = html_url.indexOf("/", start);

            String courseIdString = html_url.substring(start,end);

            return Long.parseLong(courseIdString);
        }
        return 0;
    }

    private long parseGroupId() {
        if(html_url.length() > 0 && !html_url.equals("null")) {
            String searchFor = "groups/";
            int start = html_url.indexOf(searchFor);
            if (start == -1) {
                return 0;
            }
            start += searchFor.length();
            int end = html_url.indexOf("/", start);

            String groupIdString = html_url.substring(start,end);

            return Long.parseLong(groupIdString);
        }
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(this.updated_at);
        dest.writeLong(this.id);
        dest.writeString(this.title);
        dest.writeString(this.message);
        dest.writeString(this.type);
        dest.writeString(this.context_type);
        dest.writeByte(read_state ? (byte) 1 : (byte) 0);
        dest.writeString(this.url);
        dest.writeString(this.html_url);
        dest.writeLong(this.course_id);
        dest.writeLong(this.group_id);
        dest.writeLong(this.assignment_id);
        dest.writeLong(this.message_id);
        dest.writeString(this.notification_category);
        dest.writeLong(this.conversation_id);
        dest.writeByte(isPrivate ? (byte) 1 : (byte) 0);
        dest.writeInt(this.participant_count);
        dest.writeLong(this.discussion_topic_id);
        dest.writeLong(this.announcement_id);
        dest.writeInt(this.total_root_discussion_entries);
        dest.writeByte(require_initial_post ? (byte) 1 : (byte) 0);
        dest.writeByte(user_has_posted ? (byte) 1 : (byte) 0);
        dest.writeList(this.root_discussion_entries);
        dest.writeInt(this.attempt);
        dest.writeString(this.body);
        dest.writeString(this.grade);
        dest.writeByte(grade_matches_current_submission ? (byte) 1 : (byte) 0);
        dest.writeString(this.graded_at);
        dest.writeLong(this.grader_id);
        dest.writeDouble(this.score);
        dest.writeString(this.submission_type);
        dest.writeString(this.submitted_at);
        dest.writeString(this.workflow_state);
        dest.writeByte(late ? (byte) 1 : (byte) 0);
        dest.writeString(this.preview_url);
        dest.writeList(this.submission_comments);
        dest.writeParcelable(this.canvasContext, 0);
        dest.writeParcelable(this.assignment, flags);
        dest.writeLong(this.user_id);
        dest.writeParcelable(this.user, 0);
        dest.writeInt(this.enumType == null ? -1 : this.enumType.ordinal());
        dest.writeInt(this.canvasContextType == null ? -1 : this.canvasContextType.ordinal());
        dest.writeByte(hasSetContextType ? (byte) 1 : (byte) 0);
        dest.writeLong(updatedAtDate != null ? updatedAtDate.getTime() : -1);
        dest.writeLong(gradedAtDate != null ? gradedAtDate.getTime() : -1);
        dest.writeLong(submittedAtDate != null ? submittedAtDate.getTime() : -1);
        dest.writeParcelable(this.conversation, flags);
        dest.writeByte(isChecked ? (byte) 1 : (byte) 0);
    }

    public StreamItem() {
    }

    private StreamItem(Parcel in) {
        this.updated_at = in.readString();
        this.id = in.readLong();
        this.title = in.readString();
        this.message = in.readString();
        this.type = in.readString();
        this.context_type = in.readString();
        this.read_state = in.readByte() != 0;
        this.url = in.readString();
        this.html_url = in.readString();
        this.course_id = in.readLong();
        this.group_id = in.readLong();
        this.assignment_id = in.readLong();
        this.message_id = in.readLong();
        this.notification_category = in.readString();
        this.conversation_id = in.readLong();
        this.isPrivate = in.readByte() != 0;
        this.participant_count = in.readInt();
        this.discussion_topic_id = in.readLong();
        this.announcement_id = in.readLong();
        this.total_root_discussion_entries = in.readInt();
        this.require_initial_post = in.readByte() != 0;
        this.user_has_posted = in.readByte() != 0;
        in.readList(this.root_discussion_entries, DiscussionEntry.class.getClassLoader());
        this.attempt = in.readInt();
        this.body = in.readString();
        this.grade = in.readString();
        this.grade_matches_current_submission = in.readByte() != 0;
        this.graded_at = in.readString();
        this.grader_id = in.readLong();
        this.score = in.readDouble();
        this.submission_type = in.readString();
        this.submitted_at = in.readString();
        this.workflow_state = in.readString();
        this.late = in.readByte() != 0;
        this.preview_url = in.readString();
        in.readList(this.submission_comments, SubmissionComment.class.getClassLoader());
        this.canvasContext = in.readParcelable(CanvasContext.class.getClassLoader());
        this.assignment = in.readParcelable(Assignment.class.getClassLoader());
        this.user_id = in.readLong();
        this.user = in.readParcelable(User.class.getClassLoader());
        int tmpEnumType = in.readInt();
        this.enumType = tmpEnumType == -1 ? null : Type.values()[tmpEnumType];
        int tmpCanvasContextType = in.readInt();
        this.canvasContextType = tmpCanvasContextType == -1 ? null : CanvasContext.Type.values()[tmpCanvasContextType];
        this.hasSetContextType = in.readByte() != 0;
        long tmpUpdatedAtDate = in.readLong();
        this.updatedAtDate = tmpUpdatedAtDate == -1 ? null : new Date(tmpUpdatedAtDate);
        long tmpGradedAtDate = in.readLong();
        this.gradedAtDate = tmpGradedAtDate == -1 ? null : new Date(tmpGradedAtDate);
        long tmpSubmittedAtDate = in.readLong();
        this.submittedAtDate = tmpSubmittedAtDate == -1 ? null : new Date(tmpSubmittedAtDate);
        this.conversation = in.readParcelable(Conversation.class.getClassLoader());
        this.isChecked = in.readByte() != 0;
    }

    public static Creator<StreamItem> CREATOR = new Creator<StreamItem>() {
        public StreamItem createFromParcel(Parcel source) {
            return new StreamItem(source);
        }

        public StreamItem[] newArray(int size) {
            return new StreamItem[size];
        }
    };
}

File: libs/canvas-api-2/src/main/java/com/instructure/canvasapi2/models/StreamItem.kt
/*
 * Copyright (C) 2017 - present Instructure, Inc.
 *
 *     Licensed under the Apache License, Version 2.0 (the "License");
 *     you may not use this file except in compliance with the License.
 *     You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *     Unless required by applicable law or agreed to in writing, software
 *     distributed under the License is distributed on an "AS IS" BASIS,
 *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *     See the License for the specific language governing permissions and
 *     limitations under the License.
 *
 */

package com.instructure.canvasapi2.models

import android.content.Context
import com.google.gson.annotations.SerializedName
import com.instructure.canvasapi2.R
import com.instructure.canvasapi2.utils.APIHelper
import com.instructure.canvasapi2.utils.convertScoreToLetterGrade
import com.instructure.canvasapi2.utils.toDate
import kotlinx.parcelize.IgnoredOnParcel
import kotlinx.parcelize.Parcelize
import java.util.*

@Parcelize
data class StreamItem(
        override val id: Long = -1,
        @SerializedName("updated_at")
        val updatedAt: String = "",
        @SerializedName("submitted_at")
        val submittedAt: String? = null,
        var title: String? = null,
        private var message: String? = null,
        val type: String = "",
        private val context_type: String? = null,
        // Helper method to show that the stream item has been read without having to reload all the data.
        // this method does not get the data from the server, so make sure item is actually read.
        @SerializedName("read_state")
        var isReadState: Boolean = false,
        val url: String? = null,
        @SerializedName("html_url")
        val htmlUrl: String = "",
        private var course_id: Long = -1,
        private var group_id: Long = -1,
        var assignment_id: Long = -1,

        // message type, which is not a conversation, but stream notifications
        @SerializedName("message_id")
        val messageId: Long = -1,
        @SerializedName("notification_category")
        val notificationCategory: String = "",

        // Conversation
        @SerializedName("conversation_id")
        val conversationId: Long = -1,
        @SerializedName("private")
        val isPrivate: Boolean = false,
        @SerializedName("participant_count")
        val participantCount: Int = 0,

        // discussionTopic or announcement
        private val discussion_topic_id: Long = -1,
        private val announcement_id: Long = -1,
        private val require_initial_post: Boolean = false,
        private val user_has_posted: Boolean = false,
        @SerializedName("root_discussion_entries")
        val rootDiscussionEntries: List<DiscussionEntry> = ArrayList(),

        // submission
        val attempt: Int = 0,
        val body: String? = null,
        val grade: String? = null,
        private val grade_matches_current_submission: Boolean = false,
        private val graded_at: String? = null,
        @SerializedName("grader_id")
        val graderId: Long = -1,
        val score: Double = -1.0,
        @SerializedName("submission_type")
        val submissiTonType: String? = null,
        @SerializedName("workflow_state")
        val workflowState: String = "",
        val late: Boolean = false,
        @SerializedName("preview_url")
        val previewUrl: String = "",
        @SerializedName("submission_comments")
        val submissionComments: List<SubmissionComment> = ArrayList(),
        var canvasContext: CanvasContext? = null,
        val assignment: Assignment? = null,
        @SerializedName("user_id")
        val userId: Long = -1,
        val user: User = User(),
        val excused: Boolean = false,
        @SerializedName("latest_messages")
        val latestMessages: List<Message> = ArrayList(),
) : CanvasModel<StreamItem>() {
    // We want opposite of natural sorting order of date since we want the newest one to come first
    override val comparisonDate get() = updatedDate

    val gradedDate get() = graded_at.toDate()
    val submittedDate get() = submittedAt.toDate()
    val updatedDate: Date?
        get() {
            if (getStreamItemType() == Type.CONVERSATION && latestMessages.isNotEmpty()) {
                return latestMessages
                    .filter { it.createdAt.toDate() != null }
                    .maxBy { it.createdAt.toDate()!! }
                    .createdAt.toDate()
            }
            return updatedAt.toDate()
        }

    // Helper fields
    @IgnoredOnParcel
    private var canvasContextType: CanvasContext.Type? = CanvasContext.Type.USER
    @IgnoredOnParcel
    private var hasSetContextType = false
    @IgnoredOnParcel
    var conversation: Conversation? = null
        private set
    @IgnoredOnParcel
    var isChecked: Boolean = false

    val contextType: CanvasContext.Type?
        get() {
            if (!hasSetContextType) {
                if (context_type != null && (context_type.lowercase(Locale.getDefault()) == "course" || course_id > 0)) {
                    canvasContextType = CanvasContext.Type.COURSE
                } else if (context_type != null && (context_type.lowercase(Locale.getDefault()) == "group" || group_id > 0)) {
                    canvasContextType = CanvasContext.Type.GROUP
                }
                hasSetContextType = true
            }

            return canvasContextType
        }
    val courseId: Long
        get() {
            if (contextType === CanvasContext.Type.COURSE && course_id == -1L) {
                course_id = parseCourseId()
            }
            return course_id
        }

    val groupId: Long
        get() {
            if (contextType === CanvasContext.Type.GROUP && group_id == -1L) {
                group_id = parseGroupId()
            }
            return group_id
        }
    val assignmentId: Long
        get() {
            if (contextType === CanvasContext.Type.COURSE) {
                return parseAssignmentId()
            }
            return -1L
        }
    val discussionTopicId: Long
        get() = if (discussion_topic_id == -1L) {
            announcement_id
        } else discussion_topic_id

    enum class Type {
        DISCUSSION_TOPIC, SUBMISSION, ANNOUNCEMENT, CONVERSATION, MESSAGE, CONFERENCE, COLLABORATION, COLLECTION_ITEM, UNKNOWN, NOT_SET, DISCUSSION_MENTION, DISCUSSION_ENTRY;


        companion object {
            fun isDiscussionTopic(streamItem: StreamItem): Boolean = streamItem.getStreamItemType() == DISCUSSION_TOPIC
            fun isSubmission(streamItem: StreamItem): Boolean = streamItem.getStreamItemType() == SUBMISSION
            fun isAnnouncement(streamItem: StreamItem): Boolean = streamItem.getStreamItemType() == ANNOUNCEMENT
            fun isConversation(streamItem: StreamItem): Boolean = streamItem.getStreamItemType() == CONVERSATION
            fun isMessage(streamItem: StreamItem): Boolean = streamItem.getStreamItemType() == MESSAGE
            fun isConference(streamItem: StreamItem): Boolean = streamItem.getStreamItemType() == CONFERENCE
            fun isCollaboration(streamItem: StreamItem): Boolean = streamItem.getStreamItemType() == COLLABORATION
            fun isCollectionItem(streamItem: StreamItem): Boolean = streamItem.getStreamItemType() == COLLECTION_ITEM
            fun isUnknown(streamItem: StreamItem): Boolean = streamItem.getStreamItemType() == UNKNOWN
            fun isNotSet(streamItem: StreamItem): Boolean = streamItem.getStreamItemType() == NOT_SET
        }
    }

    fun getTitle(context: Context): String? {
        if (title == null && getStreamItemType() == Type.CONVERSATION) {
            title = context.getString(R.string.Message)
        }
        return title
    }

    fun getMessage(context: Context, restrictQuantitativeData: Boolean = false, gradingScheme: List<GradingSchemeRow> = emptyList()): String? {
        if (message == null) {
            message = createMessage(context, restrictQuantitativeData, gradingScheme)
        }
        return message
    }

    fun getStreamItemType(): Type = typeFromString(type)

    private fun typeFromString(type: String): Type = when {
        type.lowercase(Locale.getDefault()) == "conversation" -> Type.CONVERSATION
        type.lowercase(Locale.getDefault()) == "submission" -> Type.SUBMISSION
        type.lowercase(Locale.getDefault()) == "discussiontopic" -> Type.DISCUSSION_TOPIC
        type.lowercase(Locale.getDefault()) == "announcement" -> Type.ANNOUNCEMENT
        type.lowercase(Locale.getDefault()) == "message" && notificationCategory.lowercase() == "discussionmention" -> Type.DISCUSSION_MENTION
        type.lowercase(Locale.getDefault()) == "message" -> Type.MESSAGE
        type.lowercase(Locale.getDefault()) == "conference" -> Type.CONFERENCE
        type.lowercase(Locale.getDefault()) == "webconference" -> Type.CONFERENCE
        type.lowercase(Locale.getDefault()) == "collaboration" -> Type.COLLABORATION
        type.lowercase(Locale.getDefault()) == "collectionitem" -> Type.COLLECTION_ITEM
        type.lowercase(Locale.getDefault()) == "discussionentry" -> Type.DISCUSSION_ENTRY
        else -> Type.UNKNOWN
    }

    fun setConversation(context: Context, conversation: Conversation?, myUserId: Long, monologueDefault: String) {
        this.conversation = conversation ?: return
        title = conversation.getMessageTitle(context, myUserId, monologueDefault).toString()
        message = createMessage(context)
    }

    fun setCanvasContextFromMap(courseMap: Map<Long, Course>, groupMap: Map<Long, Group>) {
        canvasContext = if (contextType === CanvasContext.Type.COURSE) {
            courseMap[courseId]
        } else {
            groupMap[groupId]
        }
    }

    private fun createMessage(context: Context, restrictQuantitativeData: Boolean = false, gradingScheme: List<GradingSchemeRow> = emptyList()): String? {
        when (getStreamItemType()) {
            StreamItem.Type.CONVERSATION -> {
                if (conversation == null) {
                    return context.getString(R.string.Loading)
                } else if (conversation!!.lastMessagePreview == null) {
                    return ""
                }
                return conversation!!.lastMessagePreview
            }
            StreamItem.Type.SUBMISSION -> {
                // Get comments from assignment
                var comment: String = ""
                if (submissionComments.isNotEmpty()) {
                    val lastComment = submissionComments.last().comment
                    if (lastComment != null && lastComment != "null") comment = lastComment
                }

                val displayedGrade = when {
                    excused -> context.getString(R.string.gradeExcused)
                    restrictQuantitativeData -> getGradeWhenQuantitativeDataRestricted(context, gradingScheme, score, assignment?.pointsPossible)
                    score != -1.0 -> score.toString().orEmpty()
                    else -> ""
                }

                return "$displayedGrade $comment"
            }
            StreamItem.Type.DISCUSSION_TOPIC ->
                // If it's a discussionTopic, get the last entry for the message.
                if (rootDiscussionEntries.isNotEmpty()) {
                    return rootDiscussionEntries[rootDiscussionEntries.size - 1].getMessage(context.getString(R.string.Deleted))
                }
            else -> {
            }
        }

        return if (message == null) {
            ""
        } else message
    }

    private fun getGradeWhenQuantitativeDataRestricted(context: Context, gradingScheme: List<GradingSchemeRow>, score: Double, maxScore: Double?): String {
        return if (assignment?.isGradingTypeQuantitative == true) {
            if (gradingScheme.isEmpty() || maxScore == null) {
                context.getString(R.string.gradeUpdated)
            } else {
                convertScoreToLetterGrade(score, maxScore, gradingScheme)
            }
        } else {
            grade.orEmpty()
        }
    }

    private fun parseAssignmentId(): Long {
        // Get the assignment from the url
        if (htmlUrl.isNotEmpty() && htmlUrl != "null") {
            val searchFor = "assignments/"
            var start = htmlUrl.indexOf(searchFor)
            if (start == -1) {
                return 0
            }
            start += searchFor.length
            var end = htmlUrl.indexOf("/", start)
            //in some urls the assignmentID might be the last thing so there wouldn't be a final /
            if (end == -1) {
                end = htmlUrl.length
            }
            val assignmentId = htmlUrl.substring(start, end)

            return APIHelper.expandTildeId(assignmentId).toLong()
        }
        return 0
    }

    private fun parseCourseId(): Long {
        if (htmlUrl.isNotEmpty() && htmlUrl != "null") {
            val searchFor = "courses/"
            var start = htmlUrl.indexOf(searchFor)
            if (start == -1) {
                return 0
            }
            start += searchFor.length
            val end = htmlUrl.indexOf("/", start)

            val courseIdString = htmlUrl.substring(start, end)

            return APIHelper.expandTildeId(courseIdString).toLong()
        }
        return 0
    }

    private fun parseGroupId(): Long {
        if (htmlUrl.isNotEmpty() && htmlUrl != "null") {
            val searchFor = "groups/"
            var start = htmlUrl.indexOf(searchFor)
            if (start == -1) {
                return 0
            }
            start += searchFor.length
            val end = htmlUrl.indexOf("/", start)

            val groupIdString = htmlUrl.substring(start, end)

            return APIHelper.expandTildeId(groupIdString).toLong()
        }
        return 0
    }
}
